src/compiler/emitter.ts

```diff
@@ -60,6 +68,30 @@ module ts {
             sourceMaps: sourceMapDataList
         };
 
+        function isNodeDescendentOf(node: Node, ancestor: Node): boolean {
+            while (node) {
+                if (node === ancestor) return true;
+                node = node.parent;
+            }
+            return false;
+        }
+
+        function isUniqueLocalName(name: string, container: Node): boolean {
+            for (let node = container; isNodeDescendentOf(node, container); node = node.nextContainer) {

```

Comment:
I would love if we could switch away from the .nextContainer approach, and just have .childContainers.  But that's optional and not necessary for this change.  (It would make the isUniqueLocalName simpler IMO).